tests/pdk/finfet_pdk/test_circuits.py

```diff
@@ -282,3 +283,33 @@ def test_cs_2():
     constraints = [{"constraint": "MultiConnection", "nets": ["vop"], "multiplier": 2}]
     example = build_example(name, netlist, constraints)
     run_example(example, cleanup=False, log_level=LOG_LEVEL)
+
+
+def test_charge_pump_switch():
+    name = f'ckt_{get_test_id()}'
+    netlist = circuits.charge_pump_switch(name, size=16)
```

Comment:
@stevenmburns , you can change the array size here now.

Reply:
> To reduce the placement dimensionality, this PR uses:
> 
> * Generates AlignInOrder instead of Align for large arrays
> * Generates SameTemplate

SameTemplate is fine. I think we should not add order constraints and let the placer decide the order as the order in compiler is based on the name. We do not have a method to identify the order at the compiler stage.

Reply:
duplicates are filtered now.

> > To reduce the placement dimensionality, this PR uses:
> > 
> > * Generates AlignInOrder instead of Align for large arrays
> > * Generates SameTemplate
> 
> SameTemplate is fine. I think we should not add order constraints and let the placer decide the order as the order in compiler is based on the name. We do not have a method to identify the order at the compiler stage.

duplicates are filtered now, but please consider going back to Align from AlignInOrder

Reply:
I think this depends on what is considered as an array. For example, if the array consists of instances of a single subcircuit with same connectivity, there is no point in searching different orderings (and eliminates n! redundant permutations). 

Is there any write-up on how array detection works, what is considered as array etc? or should I walk thru the code?

Reply:
![image](https://user-images.githubusercontent.com/40378093/163017860-20538fb0-ebef-4562-a53f-9ecf49b3ce78.png)
These are the kinds of arrays we handle now. We start from a high fanout net and do a BFS for matching nodes in graph. 
At the first level, from all the fanouts of the net, we find a smaller group of matching nodes. This smaller subgroup is further traversed in BFS and all the nodes should match for it to be an array and terminates when a mismatch of any element is found. 

After identification if the identified array has just one level we create a "Align" constraint. If more than one levels  we create a new hierarchy of the array along with "Align" constraint.

Reply:
@kkunal1408 , in both examples, alphanumeric ordering could indeed result in a suboptimal placement unless the ordering takes the connectivity into account. I will revert AlignInOrder to Align. However, I will change line to h_bottom to generate row-oriented layout.

Futher questions:
1. The compiler requires parameters to be the same for the repeating subcircuit. This implies that it is always possible to use SameTemplate, is this true? 
2. Does array detection create a new subcircuit (e.g. 3 resistors and two switches in the first example) and instantiate this subcircuit so that all instances have the same abstract name? I just want to make sure we can have the same template.

Reply:
Having said that, let's think if we can enhance the array detection such that it yields AlignInOrder when the connectivity is same. Else, there is a significant runtime penalty.

Reply:
> @kkunal1408 , in both examples, alphanumeric ordering could indeed result in a suboptimal placement unless the ordering takes the connectivity into account. I will revert AlignInOrder to Align. However, I will change line to h_bottom to generate row-oriented layout.
> 
> Futher questions:
> 
> 1. The compiler requires parameters to be the same for the repeating subcircuit. This implies that it is always possible to use SameTemplate, is this true?
yes

> 2. Does array detection create a new subcircuit (e.g. 3 resistors and two switches in the first example) and instantiate this subcircuit so that all instances have the same abstract name? I just want to make sure we can have the same template.

yes